quickwit-cli/src/indexing/document_indexer.rs

```diff
@@ -0,0 +1,130 @@
+
+    You should have received a copy of the GNU Affero General Public License
+    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+use crate::cli_command::IndexDataArgs;
+use tantivy::{schema::Schema, Document};
+use tokio::sync::mpsc::{Receiver, Sender};
+use tokio::task;
+use tokio::task::JoinHandle;
+
+use crate::indexing::statistics::StatisticEvent;
+
+use crate::indexing::split::Split;
+
+pub async fn index_documents(
+    args: &IndexDataArgs,
+    mut document_receiver: Receiver<String>,
+    statistic_sender: Sender<StatisticEvent>,
+) -> anyhow::Result<Vec<Split>> {
+    //TODO get schema from docBuilder
+    let schema = Schema::builder().build();
+
+    let mut split_merge_tasks = vec![];
+
+    let mut current_split = Split::create(&args, schema.clone()).await?;
+    while let Some(raw_doc) = document_receiver.recv().await {
+        let doc_size = raw_doc.as_bytes().len();
+        let parse_result = parse_document(raw_doc);
+        let doc = match parse_result {
+            Ok(doc) => {
+                statistic_sender
+                    .send(StatisticEvent::NewDocument {
+                        size_in_bytes: doc_size,
+                        error: false,
+                    })
+                    .await?;
+                current_split.metadata.num_records += 1;
+                current_split.metadata.size_in_bytes += doc_size;
+                doc
+            }
+            Err(_) => {
+                statistic_sender
+                    .send(StatisticEvent::NewDocument {
+                        size_in_bytes: doc_size,
+                        error: true,
+                    })
+                    .await?;
+                current_split.metadata.num_parsing_errors += 1;
+                continue;
+            }
+        };
+
+        current_split.add_document(doc)?;
+        if current_split.has_enough_docs() {
+            let split = std::mem::replace(
+                &mut current_split,
+                Split::create(&args, schema.clone()).await?,
+            );
+
+            let finalise_split_task = finalize_split(split, statistic_sender.clone()).await?;
+            split_merge_tasks.push(finalise_split_task);
+        }
+    }
+
+    // Finalise last split if it has docs
+    if current_split.metadata.num_records > 0 {
+        let split = std::mem::replace(
+            &mut current_split,
+            Split::create(&args, schema.clone()).await?,
+        );
+
+        let finalise_split_task = finalize_split(split, statistic_sender.clone()).await?;
+        split_merge_tasks.push(finalise_split_task);
+    }
+
+    // for all split finilisation tasks
+    let merge_results = futures::future::try_join_all(split_merge_tasks).await?;
+    //TODO let (splits, errors) = merge_results.iter().partition(|item| item.is_ok());
+
+    let splits = merge_results
+        .into_iter()
+        .filter(|result| result.is_ok())
```

Comment:
you can use `filter_map` or `flat_map` instead of the filter / map for that.
That will avoid the unwrap.

At this point I have no clue what an error means however.
What kind of error are we trying to dismiss here?